<!doctype html>

<html lang="fr">
<head>
  <meta charset="utf-8">

  <title>veqtl-mapper</title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="description" content="">
  <meta name="author" content="funpopgen">

  <link rel="stylesheet" href="https://funpopgen.github.io/veqtl-mapper/css/normalize.css">
  <link rel="stylesheet" href="https://funpopgen.github.io/veqtl-mapper/css/milligram.min.css">
  <link rel="stylesheet" href="https://funpopgen.github.io/veqtl-mapper/css/ovac-couscous.css">
  <link rel="stylesheet" href="https://funpopgen.github.io/veqtl-mapper/css/emojifyBase.min.css">
  <link rel="stylesheet" href="https://funpopgen.github.io/veqtl-mapper/css/emojify.min.css">
  <link rel="stylesheet" href="https://funpopgen.github.io/veqtl-mapper/css/emojify-emoticons.min.css">
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.4/styles/github.min.css">

  <!--[if lt IE 9]>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv.js"></script>
  <![endif]-->
</head>

<body>
    <!-- Google Tag Manager (noscript) -->
    <noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-TWFSSDL"
    height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
    <!-- End Google Tag Manager (noscript) -->
    <nav id="menu" class="panel" role="navigation">
        <h5>Table of Contents</h5>
                <ul>
                            <p class="text-muted">
                    Contents
                </p>

                <ul class="nav nav-pills nav-stacked">
                                            <li class="">
                            <a href="https://funpopgen.github.io/veqtl-mapper/">
                                Home
                            </a>
                        </li>
                                            <li class="">
                            <a href="https://funpopgen.github.io/veqtl-mapper/">
                                Contents
                            </a>
                        </li>
                                    </ul>
                            <p class="text-muted">
                    Links
                </p>

                <ul class="nav nav-pills nav-stacked">
                                            <li class="">
                            <a href="https://github.com/funpopgen/veqtl-mapper">
                                View page on github
                            </a>
                        </li>
                                            <li class="">
                            <a href="https://github.com/funpopgen/veqtl-mapper/releases/tag/1.01">
                                Download page
                            </a>
                        </li>
                                            <li class="">
                            <a href="https://doi.org/10.1093/bioinformatics/btx273">
                                Paper
                            </a>
                        </li>
                                    </ul>
            
                    </ul>
            </nav>
    <div class="push">
        <header>
            <a href="#menu" class="menu-link">&#9776;</a>
            <h1>veqtl-mapper</h1>
            <a class="github" href="http://github.com/funpopgen/veqtl-mapper" title="veqtl-mapper sur Github" target="blank">
                <svg width="80" height="80" viewBox="0 0 250 250" class="github">
                    <path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path>
                    <path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path>
                    <path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path>
                </svg>
            </a>
        </header>
        <div class="container">
                            <h1 id="veqtl-mapper">veqtl-mapper</h1>
<p>Genetic loci which are associated with the variance, rather than the mean, of a trait are known as v-eQTL. These can indicate the presence of genetic interactions, gene-environment interactions, parent of origin effects, linkage with mean effect variants, or canalisation. This package provides a tool for discovering these loci affecting molecular phenotypes. A paper accompanying this software can be found <a href="https://doi.org/10.1093/bioinformatics/btx273">here</a>.</p>
<hr />
<h2 id="contents">Contents</h2>
<ol>
<li><a href="#downloading-veqtl-mapper">Downloading veqtl-mapper</a></li>
<li><a href="#input-data-formats">Input data formats</a></li>
<li><a href="#mapping-v-eqtl">Mapping v-eQTL</a></li>
<li><a href="#mapping-variance-effects-related-to-parent-of-origin-eqtl">Mapping variance effects related to parent of origin eQTL</a></li>
<li><a href="#processing-results">Processing results</a></li>
<li><a href="#controlling-for-eqtls">Controlling for eQTLs</a></li>
<li><a href="#controlling-for-covariates">Controlling for covariates</a></li>
<li><a href="#building-from-source">Building from source</a></li>
</ol>
<hr />
<h2 id="downloading-veqtl-mapper">Downloading veqtl-mapper</h2>
<p>veqtl-mapper binaries can be downloaded from here:</p>
<p><a href="https://github.com/funpopgen/veqtl-mapper/releases/tag/1.01"><a href="https://github.com/funpopgen/veqtl-mapper/releases/tag/1.01">https://github.com/funpopgen/veqtl-mapper/releases/tag/1.01</a></a></p>
<p>The standard binary is called veqtl-mapper. Use veqtl-mapper_CentOS if there is an error referring to GLIBC.</p>
<p>veqtl-mapper requires tabix to be installed, tabix can be found here: <a href="https://github.com/samtools/htslib"><a href="https://github.com/samtools/htslib">https://github.com/samtools/htslib</a></a>.</p>
<p>Sometimes, permissions have been stripped from the binaries. If you get a &quot;Permission denied.&quot; error, this can be solved by running:</p>
<pre><code>chmod 755 veqtl-mapper</code></pre>
<p>If there are any other issues, you can try <a href="#building-from-source">building from source</a>, or contact <a href="mailto:andrew.brown@unige.ch">andrew.brown@unige.ch</a>.</p>
<hr />
<h2 id="input-data-formats">Input data formats</h2>
<p>veqtl-mapper uses the same data formats as fastQTL <a href="http://fastqtl.sourceforge.net/"><a href="http://fastqtl.sourceforge.net/">http://fastqtl.sourceforge.net/</a></a>, with the exception that the bed file should be uncompressed. To convert the fastQTL bed file run the following line:</p>
<pre><code>gunzip myphenotypefile.bed.gz</code></pre>
<hr />
<h2 id="mapping-v-eqtl">Mapping v-eQTL</h2>
<p>v-eQTL, SNPs associated with the variance of gene expression, can be caused by epistasis, gene-environment interactions, and partial linkage with eQTL (<a href="https://elifesciences.org/content/3/e01381">Brown et al., 2014</a>).</p>
<p>To run veqtl-mapper, all that is necessary is a bed file containing expression values (expression.bed) and a vcf file containing genotype values (genotype.vcf.gz, either the DS or GT field must be present).</p>
<p>veqtl-mapper has been designed so that the full analysis can be broken up into chunks, which can be submitted as cluster jobs separately. This requires two flags (if neither are present the whole analysis will be submitted), <code>--genes</code> specifies the number of genes to be analysed in each job, <code>--job-number</code> indexes the job. Therefore, if <code>--genes 10</code> is specified, <code>--job-number 1</code> will analyse genes 1-10 in the bed file, <code>--job-number 2</code> will process genes 11-20 and so on.</p>
<p>To submit a job array, with a bed file of 9,995 genes, where each job analyses 50 genes (9,995 / 50 = 199.9, so we need to run 200 jobs to cover all genes), we would submit the following command to an LSF cluster:</p>
<pre><code> bsub -o out -J"veqtl[1-200]" \
     "veqtl-mapper --bed expression.bed --vcf genotype.vcf.gz --genes 50 \
     --job-number \$LSB_JOBINDEX --out results\$LSB_JOBINDEX"</code></pre>
<p>In addition, the <code>--perm</code> option allows you to set the number of permutations used in the analysis, and optionally a seed. The <code>--normal</code> flag can also be passed to map phenotype data onto a normal distribution to remove mean-variance effects and outlier driven false positives.</p>
<p>Each results file should look like this:</p>
<p><img src="screenshots/VEQM.results.png" alt="veqtl-mapper results example" /></p>
<p>The first 5 columns indicate the gene and SNP (specified by chromosome, location, and reference and alternate alleles), then we report the Spearman correlation between the &quot;distance&quot; measure and genotype (positive indicates increased variance with the alternate allele), P value, permutation P value (not controlled for multiple testing), permutation P value controlled for multiple testing across the SNPs, and the beta approximation adjusted P value.</p>
<h2 id="mapping-variance-effects-related-to-parent-of-origin-eqtl">Mapping variance effects related to parent of origin eQTL</h2>
<p>Parent of origin eQTL occur when the effect of a genetic variant on expression depends on whether the variant is inherited from the mother or the father. Such effects can be observed as an increase in variance in the heterozygote group (<a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004508">Hoggart et al., 2014</a>). It is recommended that the genotype file only contain variants with reasonable numbers of minor allele homozygotes and heterozygotes if this analysis is run (we recommend more than 50 in both categories), as otherwise standard v-eQTL could mistakenly be discovered. The command to run this analysis is the same as the standard analysis, with the addition of the flag <code>--het</code>:</p>
<pre><code> bsub -o out -J"parent[1-200]" \
     "veqtl-mapper --bed expression.bed --vcf genotype.vcf.gz --genes 50 \
     --job-number \$LSB_JOBINDEX --het --out results\$LSB_JOBINDEX"</code></pre>
<h2 id="processing-results">Processing results</h2>
<p>Running the commands described above will produce 200 results files, results1-results200. To concatenate them together run:</p>
<pre><code> awk 'FNR&gt;1||NR==1' results* &gt; results.all</code></pre>
<p>To extract the most significant association for each gene, run the following command:</p>
<pre><code>(head -1 results.all; tail -n+2 results.all |  \
    awk '{if (best[$1]=="" || p_best[$1] &gt; $7) {best[$1] = $0; p_best[$1]=$7}} \
    END{for (var in best) print best[var]}') &gt; results.best</code></pre>
<p>The results are corrected for multiple testing across the SNPs. To correct for multiple testing across the genes as well, we can use the <a href="http://genomine.org/papers/Storey_FDR_2011.pdf">qvalue package</a> in R.</p>
<p>Within R, load the data with the following command:</p>
<pre><code>results &lt;- read.table(file = "results.best", header = T)</code></pre>
<p>A new object with adjusted P values can then be calculated:</p>
<pre><code>library(qvalue)
results &lt;- data.frame(results, Qvalues = qvalue(results$BETA)$qvalues)</code></pre>
<p>To write out this table, run:</p>
<pre><code>write.table(results, file = "results.adjusted", col.names = T, row.names = F, sep = '\t', quote = F)</code></pre>
<p>The final column in results.adjusted now contains Q values, which sets a threshold for estimating the false discovery rate while accounting for all SNPs and genes tested.</p>
<h2 id="controlling-for-eqtls">Controlling for eQTLs</h2>
<p>Variance effects can be caused by SNPs being in partial linkage disequilibrium with an eQTL. To control for these effects, you can pass a list of eQTLs to remove when mapping v-eQTLs. This list should be a whitespace separated file, one row per eQTL, with 5 columns: gene, chromosome, base pair position, reference allele and alt allele. The file is passed with the <code>--eqtl</code> flag:</p>
<pre><code> bsub -o out -J"parent[1-200]" \
 "veqtl-mapper --bed expression.bed --vcf genotype.vcf.gz --genes 50 \
 --job-number \$LSB_JOBINDEX --eqtl eqtl_file --out results\$LSB_JOBINDEX"</code></pre>
<h2 id="controlling-for-covariates">Controlling for covariates</h2>
<p>A file containing a set of covariates can also be passed. This should be a whitespace delimited file, with a header with the subject IDs and one row per individual. All covariates should be numeric. The file is passed with the <code>--cov</code> flag:</p>
<pre><code>  bsub -o out -J"parent[1-200]" \
 "veqtl-mapper --bed expression.bed --vcf genotype.vcf.gz --genes 50 \
 --job-number \$LSB_JOBINDEX --cov covariates_file --out results\$LSB_JOBINDEX"</code></pre>
<hr />
<h2 id="building-from-source">Building from source</h2>
<p>This requires a D compiler and a version of the GNU scientific library to be installed. Running the command:</p>
<pre><code>gsl-config --version</code></pre>
<p>should report the gsl version. If this fails, please follow the instructions <a href="https://www.gnu.org/software/gsl/">here</a> to install the library, or contact your system administrator.</p>
<h3 id="cloning-the-repository">Cloning the repository</h3>
<p>First, clone the repository by running:</p>
<pre><code>git clone https://github.com/funpopgen/veqtl-mapper.git</code></pre>
<p>This should create a folder called veqtl-mapper. Inside this folder is a file called makefile, in the first two lines we can specify the location of the D compiler.</p>
<h3 id="download-a-d-compiler">Download a D compiler</h3>
<p>Then, download a D compiler from either here: <a href="https://github.com/ldc-developers/ldc/releases"><a href="https://github.com/ldc-developers/ldc/releases">https://github.com/ldc-developers/ldc/releases</a></a> or here: <a href="http://dlang.org/dmd-linux.html"><a href="http://dlang.org/dmd-linux.html">http://dlang.org/dmd-linux.html</a></a> (ldc produces faster software, my experience is that dmd works better on older operating systems). Decompress the downloaded file, then edit either the first or second line of the makefile in the veqtl-mapper folder so that it contains the full path to the relevant compiler, i.e. either:</p>
<p><strong>DMD := /path/to/dmd2/linux/bin64/dmd</strong></p>
<p>or</p>
<p><strong>LDC := /path/to/ldc2-1.1.0-beta2-linux-x86_64/bin/ldc2</strong></p>
<h3 id="building-the-software">Building the software</h3>
<p>Now, to compile, run the following command:</p>
<pre><code>make test &amp;&amp; make</code></pre>
<p>The veqtl-mapper binary should be within the bin folder in the veqtl-mapper directory, running</p>
<pre><code> ./bin/veqtl-mapper --help</code></pre>
<p>should bring up the help.</p>
                    </div>
    </div>

    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
    <script src="https://funpopgen.github.io/veqtl-mapper/js/emojify.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.4/highlight.min.js"></script>
    <script src="https://funpopgen.github.io/veqtl-mapper/js/bigSlide.js"></script>
    <script src="https://funpopgen.github.io/veqtl-mapper/js/hide-on-scroll.js"></script>
    <script>
    $(document).ready(function() {
        $('.menu-link').bigSlide({
            'easyClose': true
        });
    });
    </script>

            
        
        
    
</body>
</html>
